internal/querynode/query_collection.go

```diff
@@ -140,19 +139,32 @@ func (q *queryCollection) register() {
 		return
 	}
 
-	//TODO:: can't add new vChannel to selectCase
-	q.watcherSelectCase = make([]reflect.SelectCase, 0)
+	q.watcherSharedChan = make(chan struct{}, 1024)
 	log.Debug("register tSafe watcher and init watcher select case",
 		zap.Any("collectionID", collection.ID()),
 		zap.Any("dml channels", collection.getVChannels()),
 	)
 	for _, channel := range collection.getVChannels() {
-		q.tSafeWatchers[channel] = newTSafeWatcher()
-		q.streaming.tSafeReplica.registerTSafeWatcher(channel, q.tSafeWatchers[channel])
-		q.watcherSelectCase = append(q.watcherSelectCase, reflect.SelectCase{
-			Dir:  reflect.SelectRecv,
-			Chan: reflect.ValueOf(q.tSafeWatchers[channel].watcherChan()),
-		})
+		q.addTSafeWatcher(channel)
+	}
+}
+
+func (q *queryCollection) addTSafeWatcher(vChannel Channel) {
+	q.tSafeWatchers[vChannel] = newTSafeWatcher()
+	q.streaming.tSafeReplica.registerTSafeWatcher(vChannel, q.tSafeWatchers[vChannel])
+	log.Debug("add tSafeWatcher to queryCollection",
+		zap.Any("collectionID", q.collectionID),
+		zap.Any("channel", vChannel),
+	)
+	go q.startWatcher(vChannel)
+}
+
+// TODO: add stopWatcher(), add close() to tSafeWatcher
+func (q *queryCollection) startWatcher(vChannel Channel) {
+	for {
+		// TODO: check if channel is closed
```

Comment:
This place will cause go routine leaks
suggestion:
```
	for {
                select {
                case <-ctx.Done():
			return
		case <-q.tSafeWatchers[vChannel].watcherChan():
                         q.watcherSharedChan <- struct{}{}
		// TODO: check if channel is closed
                }
	}
```

Reply:
fixed